transform.ml

```diff
@@ -1,157 +1,78 @@
+let remove_empty_jmp (seg : segment) : segment =
+  let instrs = fst seg in
+  let annots = snd seg in
+  let pred = Analysis.predecessors instrs in
+  let succ = Analysis.successors instrs in
+  let rec remove_empty_jmp pc acc_i acc_a : segment =
     let pc' = pc + 1 in
-    if pc' = Array.length prog then [prog.(pc)] else
-      match[@warning "-4"] (prog.(pc), prog.(pc')) with
-      | (Goto l1, Label l2) when l1 = l2 && List.length pred.(pc') = 1 ->
-          remove_empty_jmp (pc+2)
-      | (Label _, _) when pred.(pc) = [pc-1] && succ (pc-1) = [pc] ->
+    if pc' = Array.length instrs then
+      let acc_i, acc_a = instrs.(pc) :: acc_i, annots.(pc) :: acc_a in
+      (Array.of_list (List.rev acc_i), Array.of_list (List.rev acc_a))
+    else
+      match[@warning "-4"] instrs.(pc), instrs.(pc'), annots.(pc) with
+      | Goto l1, Label l2, a when l1 = l2 && List.length pred.(pc') = 1 ->
+        remove_empty_jmp (pc+2) acc_i acc_a
+      | Label _, _, a when pred.(pc) = [pc-1] && succ (pc-1) = [pc] ->
           (* A label is unused if the previous instruction is the only predecessor
            * unless the previous instruction jumps to it. The later can happen
            * if its a goto (then we already remove it -- see above) or if its a branch (which
            * is excluded by the second tests "succ (pc-1) = [pc]")
            * TODO: we should implement some generic api for instructions like, "Instr.is_jmp" *)
-          remove_empty_jmp pc'
-      | (_, _) ->
-          prog.(pc) :: remove_empty_jmp pc'
+        remove_empty_jmp pc' acc_i acc_a
+      | i, _, a ->
+        remove_empty_jmp pc' (i::acc_i) (a::acc_a)
   in
-  remove_empty_jmp 0
+  remove_empty_jmp 0 [] []
 
-let remove_dead_code prog entry=
-  let dead_code =
+let remove_unreachable_code (seg : segment) entry : segment =
+  let unreachable_code =
     let merge _ _ = None in
     let update _ _ = () in
-    Analysis.forward_analysis_from entry () prog merge update
+    Analysis.forward_analysis_from entry () seg merge update
   in
-  let rec remove_dead_code pc =
+  let instrs = fst seg in
+  let annots = snd seg in
+  let rec remove_unreachable_code pc acc_i acc_a : segment =
     let pc' = pc+1 in
-    if pc = Array.length prog then [] else
-      match[@warning "-4"] dead_code.(pc), prog.(pc) with
-      (* Comments are considered live, if the next instruction is live.
-       * This prevents removing comments above jump labels *)
-      | None, Comment c ->
-          if pc' = Array.length prog then [] else
-            begin match dead_code.(pc') with
-            | None -> remove_dead_code pc'
-            | Some _ -> Comment c :: remove_dead_code pc'
-            end
-      | None, _ -> remove_dead_code pc'
-      | Some _, _ -> prog.(pc) :: remove_dead_code pc'
+    if pc = Array.length instrs then
+      (Array.of_list (List.rev acc_i), Array.of_list (List.rev acc_a))
+    else
+      match[@warning "-4"] unreachable_code.(pc), instrs.(pc), annots.(pc) with
```

Comment:
You can probably drop the `[@warning "-4"]` here, as we are not matching on any head constructor anymore (it used to be `Commet`). (If it starts warning, then my guess was wrong and it needs to remain.)